kernel/kernel-impl/src/main/java/org/sakaiproject/authz/impl/SakaiSecurity.java

```diff
@@ -705,6 +724,52 @@ protected boolean checkAuthzGroups(String userId, String function, String entity
 			azgs = ref.getAuthzGroups(userId);
 		}
 
+		// need to know whether role swap is in effect, since we can't share the cache entry between sessions
+		// that are swapped and not swapped
+
+		String siteRef = null;
+		String roleswap = null;
+
+		// Actual code in DbAuthzGroupService will not roleswap if there's a user site ref in the list and
+		// it is acceptable. However we can't tell that without doing a database access, so be conservative
+		// and cache it separately as a role swap
+		// This code does not handle delegated access
+		if (azgs != null && userId != null && userId.equals(sessionManager().getCurrentSessionUserId())) {
```

Comment:
yes, I think 2 versions makes sense The version in DbAuthzGroupService is a lot longer and more complex, because it does more things. The combined version would be even more complex, because it would have to check whether to do certain things. Just read the code in isAllowed and you'll see. if you want a single piece of code you'll need to do a significant rewrite to pull out the things that are done only in DbAuthzGroupService. I'm not convinced that that's going to work out well. Try it. You'll see what I mean.

The version in SakaiSecurity does not do any database accesses as far as I can tell. The one in DbAuthzGroupService does. It also does stuff like a recursive call to isAllowed which we don't need, and which I believe does a database access.

There's one more optimization possible. Currently role swaps are stored in a session attribute ROLESWAP_PREFIX + azGroupId. It we can modify the code that sets and clears this so it also creates a session attribute IS_ROLESWAPPED if any roleswap is set, then we can bypass the whole check unless that's set. Is the worth doing now? The only real issue is checking all the code to make sure that everyone goes through SakaiSecurity to set and clear it.